src/dns.c

```diff
@@ -2191,163 +2226,244 @@ hsk_dns_name_parse(
-        return false;
-      }
+      off++;
+      *len = off;
+      return true;
+    }
+
+    // Label size can not exceed maximum
+    if (label > HSK_DNS_MAX_LABEL) {
+      *len = off;
+      return false;
+    }
+
+    // Label plus size byte can not extend full name past maximum
+    if (label + 1 + noff > HSK_DNS_MAX_NAME) {
+      *len = off;
+      return false;
+    }
+
+    // Label compression
+    if (cmp) {
+      // Check label map for remainder of name
+      size_t p = (size_t)hsk_map_get(&cmp->map, &name[noff]);
+
+      // Add remainder of name to map if it's not there already
+      if (p == 0) {
+        size_t o = (size_t)(&data[off] - cmp->msg);
 
-      if (data) {
-        if (off + 1 + size > data_len) {
+        if (o < (2 << 13))
+          hsk_map_set(&cmp->map, &name[noff], (void *)o);
+      } else {
+        // If the string was already in the map,
+        // point to it instead of repeating the string.
+        // Ensure enough room for 2-byte pointer, this should only
+        // be an issue if we are pointing to a 1-byte label...!
+        if (2 + noff > HSK_DNS_MAX_NAME) {
           *len = off;
           return false;
         }
-        data[off] = size;
-      }
 
-      if (cmp) {
-        char *sub = (char *)&name[begin];
-        if (strcmp(sub, ".") != 0) {
-          size_t p = (size_t)hsk_map_get(&cmp->map, sub);
-          if (p == 0) {
-            size_t o = (size_t)(&data[off] - cmp->msg);
-            if (o < (2 << 13))
-              hsk_map_set(&cmp->map, sub, (void *)o);
-          } else {
-            if (ptr == -1) {
-              ptr = p;
-              pos = off;
-              i += strlen(s);
-              break;
-            }
-          }
+        if (data) {
+          // Encode and write pointer
+          p ^= 0xc000;
+          data[off] = (p >> 8) & 0xff;
+          data[off + 1] = p & 0xff;
         }
+
+        // Done: pointers are always last
+        off += 2;
+        *len = off;
+        return true;
       }
+    }
 
-      off += 1;
+    // Compression is off, or the label is new: write it raw
+    if (data)
+      memcpy(&data[off], &name[noff], label + 1);
 
-      if (data) {
-        int j;
-        for (j = begin; j < i; j++) {
-          char ch = name[j];
+    off += label + 1;
+    noff += label + 1;
+  }
+}
 
-          // 0xff -> NUL
-          if (ch == -1)
-            ch = '\0';
+bool
+hsk_dns_name_to_string(const uint8_t *name, char *namestr) {
+  // Offset of `name` array being read
+  int off = 0;
+  // Offset of output `namestr` string being written
+  int noff = 0;
 
-          // 0xfe -> .
-          if (ch == -2)
-            ch = '.';
+  for (;;) {
+    uint8_t label = name[off];
 
-          data[off++] = ch;
-        }
+    if (label == 0x00) {
+      if (noff == 0) {
+        // Only one label and it's root
+        namestr[noff++] = '.';
+        namestr[noff] = '\0';
       } else {
-        off += size;
+        namestr[noff] = '\0';
```

Comment:
Do we not add last `.` ? `test.` or `test`, I am not sure what convention is that for names. (FQDN?)

Reply:
We do, on line 2381. Right here we're just adding the dot in the edge case that the entire name is just `"."` with no other labels